# SENTIMENT ANALYSIS

to import the important python libraries

In [234]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn as sk

import warnings
warnings.filterwarnings('ignore')

To get the datasset from the naan mudhalvan portal

In [235]:
sentiment=pd.read_csv('/content/Tweets.csv')
sentiment.fillna(0)

tweet_id airline_sentiment  airline_sentiment_confidence  \
0     570306133677760513           neutral                        1.0000   
1     570301130888122368          positive                        0.3486   
2     570301083672813571           neutral                        0.6837   
3     570301031407624196          negative                        1.0000   
4     570300817074462722          negative                        1.0000   
...                  ...               ...                           ...   
9150  570109961511895040          negative                        1.0000   
9151  570109219078152193          negative                        1.0000   
9152  570108987284197376          negative                        1.0000   
9153  570108377029685248          negative                        1.0000   
9154  570106579879645185          negative                        1.0000   

              negativereason  negativereason_confidence         airline  \
0                          0                     0.0000  Virgin America   
1                          0                     0.0000  Virgin America   
2                          0                     0.0000  Virgin America   
3                 Bad Flight                     0.7033  Virgin America   
4                 Can't Tell                     1.0000  Virgin America   
...                      ...                        ...             ...   
9150             Late Flight                     1.0000      US Airways   
9151              Can't Tell                     0.3474      US Airways   
9152  Customer Service Issue                     0.3570      US Airways   
9153             Late Flight                     0.6405      US Airways   
9154            Lost Luggage                     0.7128      US Airways   

     airline_sentiment_gold           name negativereason_gold  retweet_count  \
0                         0        cairdin                   0              0   
1                         0       jnardino                   0              0   
2                         0     yvonnalynn                   0              0   
3                         0       jnardino                   0              0   
4                         0       jnardino                   0              0   
...                     ...            ...                 ...            ...   
9150                      0        medrisy                   0              0   
9151                      0     CicilyGray                   0              0   
9152                      0      jpkansara                   0              0   
9153                      0  jakenemmasmom                   0              0   
9154                      0   mydulcebella                   0              0   

                                                   text tweet_coord  \
0                   @VirginAmerica What @dhepburn said.           0   
1     @VirginAmerica plus you've added commercials t...           0   
2     @VirginAmerica I didn't today... Must mean I n...           0   
3     @VirginAmerica it's really aggressive to blast...           0   
4     @VirginAmerica and it's a really big bad thing...           0   
...                                                 ...         ...   
9150  @USAirways is utterly unreliable! They didn't ...           0   
9151  @USAirways yes but again valuable time has bee...           0   
9152  @USAirways I don't like being lied to. No bad ...           0   
9153  @USAirways yes it landed at 11:40am. Should no...           0   
9154                                   @USAirways the d           0   

                  tweet_created   tweet_location               user_timezone  
0     2015-02-24 11:35:52 -0800                0  Eastern Time (US & Canada)  
1     2015-02-24 11:15:59 -0800                0  Pacific Time (US & Canada)  
2     2015-02-24 11:15:48 -0800        Lets Play  Central Time (US & Canada)  
3     2015-02-24 11:15:36 -0800                0  Pacific Ti

To get the head and tail of the dataset

In [236]:
sentiment.head(4)

tweet_id airline_sentiment  airline_sentiment_confidence  \
0  570306133677760513           neutral                        1.0000   
1  570301130888122368          positive                        0.3486   
2  570301083672813571           neutral                        0.6837   
3  570301031407624196          negative                        1.0000   

  negativereason  negativereason_confidence         airline  \
0            NaN                        NaN  Virgin America   
1            NaN                     0.0000  Virgin America   
2            NaN                        NaN  Virgin America   
3     Bad Flight                     0.7033  Virgin America   

  airline_sentiment_gold        name negativereason_gold  retweet_count  \
0                    NaN     cairdin                 NaN              0   
1                    NaN    jnardino                 NaN              0   
2                    NaN  yvonnalynn                 NaN              0   
3                    NaN    jnardino                 NaN              0   

                                                text tweet_coord  \
0                @VirginAmerica What @dhepburn said.         NaN   
1  @VirginAmerica plus you've added commercials t...         NaN   
2  @VirginAmerica I didn't today... Must mean I n...         NaN   
3  @VirginAmerica it's really aggressive to blast...         NaN   

               tweet_created tweet_location               user_timezone  
0  2015-02-24 11:35:52 -0800            NaN  Eastern Time (US & Canada)  
1  2015-02-24 11:15:59 -0800            NaN  Pacific Time (US & Canada)  
2  2015-02-24 11:15:48 -0800      Lets Play  Central Time (US & Canada)  
3  2015-02-24 11:15:36 -0800            NaN  Pacific Time (US & Canada)

In [237]:
sentiment.tail(4)

tweet_id airline_sentiment  airline_sentiment_confidence  \
9151  570109219078152193          negative                           1.0   
9152  570108987284197376          negative                           1.0   
9153  570108377029685248          negative                           1.0   
9154  570106579879645185          negative                           1.0   

              negativereason  negativereason_confidence     airline  \
9151              Can't Tell                     0.3474  US Airways   
9152  Customer Service Issue                     0.3570  US Airways   
9153             Late Flight                     0.6405  US Airways   
9154            Lost Luggage                     0.7128  US Airways   

     airline_sentiment_gold           name negativereason_gold  retweet_count  \
9151                    NaN     CicilyGray                 NaN              0   
9152                    NaN      jpkansara                 NaN              0   
9153                    NaN  jakenemmasmom                 NaN              0   
9154                    NaN   mydulcebella                 NaN              0   

                                                   text tweet_coord  \
9151  @USAirways yes but again valuable time has bee...         NaN   
9152  @USAirways I don't like being lied to. No bad ...         NaN   
9153  @USAirways yes it landed at 11:40am. Should no...         NaN   
9154                                   @USAirways the d         NaN   

                  tweet_created   tweet_location user_timezone  
9151  2015-02-23 22:33:24 -0800              NaN           NaN  
9152  2015-02-23 22:32:29 -0800  Washington,D.C.           NaN  
9153  2015-02-23 22:30:03 -0800              NaN           NaN  
9154                        NaN              NaN           NaN

To elpore the dataset using python

In [238]:
sentiment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9155 entries, 0 to 9154
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   tweet_id                      9155 non-null   int64  
 1   airline_sentiment             9155 non-null   object 
 2   airline_sentiment_confidence  9155 non-null   float64
 3   negativereason                5084 non-null   object 
 4   negativereason_confidence     6007 non-null   float64
 5   airline                       9155 non-null   object 
 6   airline_sentiment_gold        21 non-null     object 
 7   name                          9155 non-null   object 
 8   negativereason_gold           17 non-null     object 
 9   retweet_count                 9155 non-null   int64  
 10  text                          9155 non-null   object 
 11  tweet_coord                   689 non-null    object 
 12  tweet_created                 9154 non-null   object 
 13  twe

To describe the dataset using python

In [239]:
sentiment.describe()

tweet_id  airline_sentiment_confidence  negativereason_confidence  \
count  9.155000e+03                   9155.000000                6007.000000   
mean   5.690615e+17                      0.888037                   0.607716   
std    7.845396e+14                      0.169241                   0.339856   
min    5.675883e+17                      0.335000                   0.000000   
25%    5.684299e+17                      0.683200                   0.352000   
50%    5.691657e+17                      1.000000                   0.666300   
75%    5.697061e+17                      1.000000                   1.000000   
max    5.703106e+17                      1.000000                   1.000000   

       retweet_count  
count    9155.000000  
mean        0.093392  
std         0.652314  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max        31.000000

To find the all duplicated values present in the dataset

In [240]:
sentiment[sentiment.duplicated()]

Empty DataFrame
Columns: [tweet_id, airline_sentiment, airline_sentiment_confidence, negativereason, negativereason_confidence, airline, airline_sentiment_gold, name, negativereason_gold, retweet_count, text, tweet_coord, tweet_created, tweet_location, user_timezone]
Index: []

There is no duplicated values present in the dataset

To identify the shape of the dataset

In [241]:
sentiment.shape

(9155, 15)

The above dataet contains 9155 rows and 15 columns in the dataset

To find the datatypes that are present in the dataset

In [242]:
sentiment.dtypes

tweet_id                          int64
airline_sentiment                object
airline_sentiment_confidence    float64
negativereason                   object
negativereason_confidence       float64
airline                          object
airline_sentiment_gold           object
name                             object
negativereason_gold              object
retweet_count                     int64
text                             object
tweet_coord                      object
tweet_created                    object
tweet_location                   object
user_timezone                    object
dtype: object

The above dataset contains the int,object,float

To display the columns present in the dataset

In [243]:
sentiment.columns

Index(['tweet_id', 'airline_sentiment', 'airline_sentiment_confidence',
       'negativereason', 'negativereason_confidence', 'airline',
       'airline_sentiment_gold', 'name', 'negativereason_gold',
       'retweet_count', 'text', 'tweet_coord', 'tweet_created',
       'tweet_location', 'user_timezone'],
      dtype='object')

To describe the all the columns present in the dataset

1.**TWEETS ID**

In [244]:
sentiment.tweet_id

0       570306133677760513
1       570301130888122368
2       570301083672813571
3       570301031407624196
4       570300817074462722
               ...        
9150    570109961511895040
9151    570109219078152193
9152    570108987284197376
9153    570108377029685248
9154    570106579879645185
Name: tweet_id, Length: 9155, dtype: int64

**2.AIRLINE SENTIMENT**

In [245]:
sentiment.airline_sentiment

0        neutral
1       positive
2        neutral
3       negative
4       negative
          ...   
9150    negative
9151    negative
9152    negative
9153    negative
9154    negative
Name: airline_sentiment, Length: 9155, dtype: object

In [248]:
sentiment['airline_sentiment'].value_counts()

Series([], Name: airline_sentiment, dtype: int64)

In [246]:
sentiment['airline_sentiment']=sentiment['airline_sentiment'].replace(to_replace=['negative','neutral','postive'],value=[1,0,2],inplace=True)

**2.AIRLINE SENTIMENT CONFIDENCE**

In [247]:
sentiment.airline_sentiment_confidence

0       1.0000
1       0.3486
2       0.6837
3       1.0000
4       1.0000
         ...  
9150    1.0000
9151    1.0000
9152    1.0000
9153    1.0000
9154    1.0000
Name: airline_sentiment_confidence, Length: 9155, dtype: float64

**3.NEGATIVE REASONS**

In [250]:
sentiment.negativereason

0                          NaN
1                          NaN
2                          NaN
3                   Bad Flight
4                   Can't Tell
                 ...          
9150               Late Flight
9151                Can't Tell
9152    Customer Service Issue
9153               Late Flight
9154              Lost Luggage
Name: negativereason, Length: 9155, dtype: object

In [251]:
sentiment['negativereason'].value_counts()

Customer Service Issue         1382
Late Flight                     982
Can't Tell                      763
Lost Luggage                    437
Cancelled Flight                418
Bad Flight                      392
Flight Booking Problems         285
Flight Attendant Complaints     277
longlines                        97
Damaged Luggage                  51
Name: negativereason, dtype: int64

In [252]:
sentiment['negativereason'].replace(to_replace=['Customer Service Issue','Late Flight','Lost Luggage','Cancelled Flight','Bad Flight','Flight Booking Problems','Flight Attendant Complaints','longlines','Damaged Luggage'],value=[1,2,3,4,5,6,7,8,9],inplace=True)

In [253]:
sentiment['negativereason'].value_counts()

1             1382
2              982
Can't Tell     763
3              437
4              418
5              392
6              285
7              277
8               97
9               51
Name: negativereason, dtype: int64

In [254]:
sentiment.drop(sentiment.index[sentiment['negativereason'] == "Can't Tell"], inplace = True)

In [286]:
sentiment['negativereason']=sentiment['negativereason'].value_counts()

**4.NEGATIVE RESON CONFIDENCE**

In [256]:
sentiment.negativereason_confidence

0          NaN
1       0.0000
2          NaN
3       0.7033
6       0.0000
         ...  
9149    0.0000
9150    1.0000
9152    0.3570
9153    0.6405
9154    0.7128
Name: negativereason_confidence, Length: 8392, dtype: float64

**5.AIRLINE**

In [257]:
sentiment.airline

0       Virgin America
1       Virgin America
2       Virgin America
3       Virgin America
6       Virgin America
             ...      
9149        US Airways
9150        US Airways
9152        US Airways
9153        US Airways
9154        US Airways
Name: airline, Length: 8392, dtype: object

In [258]:
sentiment['airline'].value_counts()

United            3443
Southwest         2261
Delta             2035
Virgin America     482
US Airways         171
Name: airline, dtype: int64

In [259]:
sentiment['airline'].replace(to_replace=['United','Southwest','Delta','Virgin America','US Airways'],value=[1,2,3,4,5],inplace=True)

In [260]:
sentiment['airline'].value_counts()

1    3443
2    2261
3    2035
4     482
5     171
Name: airline, dtype: int64

**6.AIRLINE SEMTIMENT GOLD**

In [261]:
sentiment.airline_sentiment_gold

0       NaN
1       NaN
2       NaN
3       NaN
6       NaN
       ... 
9149    NaN
9150    NaN
9152    NaN
9153    NaN
9154    NaN
Name: airline_sentiment_gold, Length: 8392, dtype: object

In [262]:
del sentiment['airline_sentiment_gold']

**7.NAME**

In [263]:
sentiment.name

0             cairdin
1            jnardino
2          yvonnalynn
3            jnardino
6          cjmcginnis
            ...      
9149     mydulcebella
9150          medrisy
9152        jpkansara
9153    jakenemmasmom
9154     mydulcebella
Name: name, Length: 8392, dtype: object

In [264]:
del sentiment['name']

8.NEGATIVE REASON GOLD

In [265]:
sentiment['negativereason_gold'].value_counts()

Customer Service Issue                      4
Cancelled Flight                            3
Late Flight                                 2
Cancelled Flight\nCustomer Service Issue    2
Late Flight\nFlight Attendant Complaints    1
Late Flight\nLost Luggage                   1
Bad Flight                                  1
Lost Luggage\nDamaged Luggage               1
Name: negativereason_gold, dtype: int64

In [266]:
sentiment['negativereason_gold'].replace(to_replace=['Customer Service Issue','Cancelled Flight','Late Flight','Cancelled Flight\nCustomer Service Issue','Late Flight\nFlight Attendant Complaints','Late Flight\nLost Luggage','Bad Flight','Lost Luggage\nDamaged Luggage'],value=[1,2,3,4,5,6,7,8],inplace=True)

In [267]:
sentiment['negativereason_gold'].value_counts()

1.0    4
2.0    3
3.0    2
4.0    2
5.0    1
6.0    1
7.0    1
8.0    1
Name: negativereason_gold, dtype: int64

**9.TEXT**

In [268]:
text=pd.DataFrame(sentiment['text'])

In [269]:
from collections import Counter
total_counts=Counter()
for i in range(len(text)):
  for word in text.values[i][0].split(" "):
    total_counts[word]+=1

print('total words in the dataset: ',len(total_counts))

total words in the dataset:  21091


In [270]:
vocabulary=sorted(total_counts,key=total_counts.get,reverse=True)
print(vocabulary[:60])

['to', 'the', '@united', 'I', 'a', '@SouthwestAir', 'for', 'on', 'and', '@JetBlue', 'you', 'my', 'flight', 'in', 'is', '', 'of', 'was', 'me', 'with', 'your', 'at', 'have', 'it', 'that', 'from', 'not', 'be', 'get', 'but', 'this', 'are', 'no', 'Cancelled', 'an', 'can', 'we', '@VirginAmerica', 'just', 'so', '-', 'do', "I'm", 'customer', 'will', 'service', 'been', 'Flight', '&amp;', 'out', 'now', 'time', 'they', 'when', 'our', 'flights', 'up', '2', 'would', 'if']


In [271]:
vocab_size=len(vocabulary)
word_1={}
for i, word in enumerate(vocabulary):
  word_1[word]=i

In [272]:
def text_to_vector(text):
  word_vector=np.zeros(vocabulary_size)
  for word in text.split(" "):
    if word_1.get(word) is None:
      continue
    else:
      word_vector[word_1.get(word)]+=1
  return np.array(word_vector)

In [273]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [274]:
vectorizer=TfidfVectorizer()

In [275]:
vectors=vectorizer.fit_transform(sentiment['text'])

In [276]:
vectors.shape

(8392, 11140)

In [277]:
features =vectors

10.TWEET CREATED

In [278]:
sentiment['tweet_created'].value_counts()

2015-02-23 15:25:46 -0800    3
2015-02-23 06:57:24 -0800    3
2015-02-23 14:26:53 -0800    2
2015-02-24 06:04:03 -0800    2
2015-02-20 16:18:35 -0800    2
                            ..
2015-02-20 02:17:18 -0800    1
2015-02-20 02:17:21 -0800    1
2015-02-20 02:18:48 -0800    1
2015-02-20 02:34:28 -0800    1
2015-02-23 22:30:03 -0800    1
Name: tweet_created, Length: 8315, dtype: int64

In [279]:
import datetime as dt

In [280]:
sentiment['tweet_created'] = pd.to_datetime(sentiment['tweet_created'])

In [281]:
day=[]
month=[]
year=[]
hour=[]

In [282]:
for i in sentiment['tweet_created']:

    month.append(i.month)
    day.append(i.day)
    hour.append(i.hour)
    year.append(i.year)

In [283]:
sentiment['month']=month
sentiment['day']=day
sentiment['hour']=hour

In [284]:
del sentiment['tweet_coord']
del sentiment['tweet_location']
del sentiment['user_timezone']
del sentiment['tweet_created']

In [311]:
x=sentiment.drop('airline_sentiment',axis=1)
y=sentiment['airline_sentiment']
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y)

In [312]:
x_train

tweet_id  airline_sentiment_confidence  negativereason  \
5224  569250176369471489                        1.0000             NaN   
8209  568604311598469121                        1.0000             NaN   
5444  569008329038524417                        1.0000             NaN   
8537  568181761114546176                        1.0000             NaN   
3934  567946616792952832                        1.0000             NaN   
...                  ...                           ...             ...   
5810  568581908075933696                        0.9658             NaN   
1537  569676535973744640                        0.6889             NaN   
657   570222630525014017                        1.0000             NaN   
7312  569672477967835136                        1.0000             NaN   
8594  568132912761208832                        1.0000             NaN   

      negativereason_confidence  airline  negativereason_gold  retweet_count  \
5224                     0.6566        2                  NaN              0   
8209                     0.3514        3                  NaN              2   
5444                        NaN        2                  NaN              0   
8537                     0.3554        3                  NaN              0   
3934                     0.6863        1                  NaN              0   
...                         ...      ...                  ...            ...   
5810                     0.6623        2                  7.0              1   
1537                     0.3549        1                  NaN              0   
657                         NaN        1                  NaN              0   
7312                     1.0000        3                  NaN              0   
8594                        NaN        3                  NaN              0   

                                                   text  month   day  hour  
5224  @SouthwestAir Booked 6mos early, paid for Earl...    2.0  21.0  13.0  
8209  @JetBlue what's going on with flight 669 and h...    2.0  19.0  18.0  
5444  @SouthwestAir Hey there, can you follow us so ...    2.0  20.0  21.0  
8537  @JetBlue u now have my gears grinding. The JFK...    2.0  18.0  14.0  
3934  @united  treats service members like crap neve...    2.0  17.0  23.0  
...                                                 ...    ...   ...   ...  
5810  @SouthwestAir please do something about the sp...    2.0  19.0  17.0  
1537  @united good try but @SouthwestAir got her her...    2.0  22.0  17.0  
657   @united I know this is probably a no but is th...    2.0  24.0   6.0  
7312  @JetBlue Thanks. So the delay was the result o...    2.0  22.0  17.0  
8594  @JetBlue When I checked yesterday it looked li...    2.0  18.0  11.0  

[6294 rows x 11 columns]

In [313]:
y_train

5224    None
8209    None
5444    None
8537    None
3934    None
        ... 
5810    None
1537    None
657     None
7312    None
8594    None
Name: airline_sentiment, Length: 6294, dtype: object

In [314]:
x_test

tweet_id  airline_sentiment_confidence  negativereason  \
8944  567728800864010242                        1.0000             NaN   
8315  568491434397073408                        1.0000             NaN   
6301  568072958821990400                        1.0000             NaN   
4880  569669872050765824                        1.0000             NaN   
7734  569299428068233217                        1.0000             NaN   
...                  ...                           ...             ...   
6281  568084748020649984                        1.0000             NaN   
7807  569235156717989888                        1.0000             NaN   
2305  569264951401054208                        1.0000             NaN   
4298  567729246198837248                        0.6702             NaN   
3334  568516490405679104                        1.0000             NaN   

      negativereason_confidence  airline  negativereason_gold  retweet_count  \
8944                     1.0000        3                  NaN              0   
8315                     1.0000        3                  NaN              0   
6301                        NaN        2                  NaN              0   
4880                        NaN        2                  NaN              0   
7734                        NaN        3                  NaN              0   
...                         ...      ...                  ...            ...   
6281                        NaN        2                  NaN              0   
7807                        NaN        3                  NaN              0   
2305                     0.6633        1                  NaN              0   
4298                        NaN        1                  NaN              0   
3334                        NaN        1                  NaN              0   

                                                   text  month   day  hour  
8944  @JetBlue busy due to storm but customer servic...    2.0  17.0   8.0  
8315  @JetBlue You mean take more of my time to writ...    2.0  19.0  11.0  
6301  @SouthwestAir pleasantly surprised to be board...    2.0  18.0   7.0  
4880                                @SouthwestAir gimme    2.0  22.0  17.0  
7734                  @JetBlue that selfie was extreme.    2.0  21.0  16.0  
...                                                 ...    ...   ...   ...  
6281  @SouthwestAir my birthday is February 24th whi...    2.0  18.0   8.0  
7807                      @JetBlue Using JetBlue miles?    2.0  21.0  12.0  
2305  @united 2/2 Instead, had to wait on plane from...    2.0  21.0  14.0  
4298  @united Filled in the form you sent the link t...    2.0  17.0   8.0  
3334  @united @perfectomobile I have to admit that n...    2.0  19.0  13.0  

[2098 rows x 11 columns]

In [315]:
y_test

8944    None
8315    None
6301    None
4880    None
7734    None
        ... 
6281    None
7807    None
2305    None
4298    None
3334    None
Name: airline_sentiment, Length: 2098, dtype: object

In [316]:
from sklearn.feature_extraction.text import CountVectorizer

In [317]:
cv=CountVectorizer()
features=cv.fit_transform(x_train['text'])
features

<6294x9392 sparse matrix of type '<class 'numpy.int64'>'
	with 97864 stored elements in Compressed Sparse Row format>

In [318]:
cv=CountVectorizer()
features_1=cv.fit_transform(x_test['text'])
features_1

<2098x4962 sparse matrix of type '<class 'numpy.int64'>'
	with 32983 stored elements in Compressed Sparse Row format>

In [319]:
from sklearn.linear_model import LinearRegression
ran=LinearRegression()

In [320]:
ran.fit(features,y_train)

LinearRegression()

In [332]:
y_pred=ran.predict(features)

In [333]:
y_pred

array([nan, nan, nan, ..., nan, nan, nan])

The sentiment analysis will be predicted successfully using python